### PR TITLE
feat(potree): data source initial

### PR DESCRIPTION
**Pull request by belom88**

### CODE REVIEW

File: modules/potree/src/lib/potree-node-source.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  url: string;
+  data: string | Blob;
+  props: PotreeNodesSourceProps;
+  metadata: PotreeMetadata | null = null;
+  root: POTreeNode | null = null;
+  isReady = false;
+
+  private initPromise: Promise<void> | null = null;
+
+  get isSupported(): boolean {
```
Comment (ibgreen) [inline modules/potree/src/lib/potree-node-source.ts:32]: No need for getters here. I think it is fine to call `isSupported()` as a function.
```suggestion
  isSupported(): boolean {
```

File: modules/potree/src/lib/potree-node-source.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    // TODO - add options here
+  };
+};
+
+/**
+ * A Potree data source
+ * @note Point cloud nodes tile source
+ */
+export class PotreeNodesSource extends DataSource {
+  url: string;
```
Comment (ibgreen) [inline modules/potree/src/lib/potree-node-source.ts:23]: Add TSDOc comment on each prop.

File: modules/potree/src/lib/potree-node-source.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    super(props);
+    this.props = props;
+    this.data = data;
+    this.url = typeof data === 'string' ? resolvePath(data) : '';
+
+    this.initPromise = this.init();
+  }
+
+  async init() {
+    this.metadata = await load(`${this.url}/cloud.js`, PotreeLoader);
```
Comment (ibgreen) [inline modules/potree/src/lib/potree-node-source.ts:65]: Is it always called `cloud.js`? Is the URL just supposed to be a directory? Maybe specify in a comment here and on the prop.

To make it easier for the user we may have logic that accepts both a filename with `cloud.js` at the end and just a directory name.

File: modules/potree/src/lib/potree-node-source.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    await this.loadHierarchy();
+    this.isReady = true;
+  }
+
+  async loadNodeContent(path: number[]): Promise<MeshGeometry | null> {
```
Comment (ibgreen) [inline modules/potree/src/lib/potree-node-source.ts:70]: A `number[]` array is very raw, doesn't give much explanation or context. I wonder if a type for the path might better help users understand how to use this API.

```ts
type NodePath = {
  /** array of numbers between 0-7 specifying successive octree divisions. */
  path: number[];
};
```

Comment (ibgreen) [inline modules/potree/src/lib/potree-node-source.ts:70]: Short TSDoc on every method?

File: modules/potree/src/lib/potree-node-source.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+  async loadHierarchy(): Promise<void> {
+    await this.initPromise;
+    this.root = await load(
+      `${this.url}/${this.metadata?.octreeDir}/r/r.hrc`,
+      PotreeHierarchyChunkLoader
+    );
+  }
+
+  async getNodeAvailability(path: number[]): Promise<boolean> {
```
Comment (ibgreen) [inline modules/potree/src/lib/potree-node-source.ts:95]: Nit: Playing with the naming... not sure what feels best.
```suggestion
  async isNodeAvailable(path: number[]): Promise<boolean> {
```

File: modules/potree/src/types/potree-metadata.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  lz: number;
+  /** Max X */
+  ux: number;
+  /** Max Y */
+  uy: number;
+  /** Max Z */
+  uz: number;
+}
+
+export type PotreeAttribute =
```
Comment (ibgreen) [inline modules/potree/src/types/potree-metadata.ts:17]: Love the rich types. Add a TSDoc line in front of each type if possible

**Review by ibgreen — APPROVED** [review]
We are off to a nice start!

File: modules/potree/src/types/potree-metadata.ts (hunk 6 above)
Comment (belom88) [reply modules/potree/src/types/potree-metadata.ts:17]: Done

File: modules/potree/src/lib/potree-node-source.ts (hunk 1 above)
Comment (belom88) [reply modules/potree/src/lib/potree-node-source.ts:32]: Done

File: modules/potree/src/lib/potree-node-source.ts (hunk 2 above)
Comment (belom88) [reply modules/potree/src/lib/potree-node-source.ts:23]: Done

File: modules/potree/src/lib/potree-node-source.ts (hunk 3 above)
Comment (belom88) [reply modules/potree/src/lib/potree-node-source.ts:65]: Done

File: modules/potree/src/lib/potree-node-source.ts (hunk 4 above)
Comment (belom88) [reply modules/potree/src/lib/potree-node-source.ts:70]: Added a TSDoc comment

Comment (belom88) [reply modules/potree/src/lib/potree-node-source.ts:70]: Done

File: modules/potree/src/lib/potree-node-source.ts (hunk 5 above)
Comment (belom88) [reply modules/potree/src/lib/potree-node-source.ts:95]: Done

**Pull request merged**